bids/layout/tests/test_layout_on_examples.py

```diff
@@ -0,0 +1,37 @@
+""" Tests runs layout on bids examples and make sure all files are caught"""
+
+from os.path import join
+
+import pytest
+
+from bids.layout import BIDSLayout
+from bids.tests import get_test_data_path
+
+# Values for the number of file got from a:
+#
+#   find ds_name -type f | wc -l
+#
+@pytest.mark.parametrize(
+    "dataset, nb_files",
+    [
+        ("micr_SEM", 12),
+        ("micr_SPIM", 22),
+        ("asl001", 8),
+        ("asl002", 10),
+        ("asl003", 10),
+        ("asl004", 12),
+        ("asl005", 10),
+        ("pet001", 12),
+        ("pet002", 20),
+        ("pet003", 9),
+        ("pet004", 10),
+        ("pet005", 14),
+        ("qmri_qsm", 8),
+        ("qmri_vfa", 17),
+    ],
+)
+def test_index_metadata(dataset, nb_files):
+    ds = join(get_test_data_path(), "bids-examples", dataset)
+    layout = BIDSLayout(ds, derivatives=True)
+    files = layout.get()
+    assert len(files) == nb_files
```

Comment:
The test assumes that the length of a query for all the files of a dataset (including derivatives) in a layout should return the same value`find ds_name -type f | wc -l`.

So far only tried on some of the datasets that are in bids-examples and are "vanilla bids" and in some cases have valid derivatives (some qMRI).

FYI this fails locally.

```bash
 short test summary info ======================================================
FAILED bids/layout/tests/test_layout_on_examples.py::test_index_metadata[micr_SEM-12] - AssertionError: assert 6 == 12
FAILED bids/layout/tests/test_layout_on_examples.py::test_index_metadata[micr_SPIM-22] - AssertionError: assert 4 == 22
FAILED bids/layout/tests/test_layout_on_examples.py::test_index_metadata[qmri_qsm-8] - AssertionError: assert 5 == 8
```

On inspection the samples TSV and JSON in the root are not picked up but neiter are the microscopy data file proper so there may be another problem somewhere else... Any thought?

@mariehbourget : I am using your config and I can't seem to be able to query the micr files...

Reply:
OK also some failure on some qMRI datasets... :-/

Before I add more datasets to parametrize this on, does this look like a sensible test to have in general? It may break when bids-examples is changed but it seems a cheap way to make sure pybids has fewer blind spots.

Bringing bids-examples on board, could also be a way to start implementing a test of the default path patterns (see [here](https://github.com/bids-standard/pybids/pull/605#issuecomment-626344752) ) until we use the schema generate those (and even then we may want to keep the test) .

Reply:
Hey @Remi-Gau thank you so much for working on this! I see that vfa and qsm are failing, would be happy to answer questions. 

For vfa, there's no fixed number of input. The qsm example on bids-example focuses more on the derivative (chimap) than it does on the raw data found in the anat folder. So for qsm, we can tweak the example for it to behave properly. 

I am not sure about the origin of these errors, but I am here for questions.

Reply:
> So for qsm, we can tweak the example for it to behave properly.

I think I would rather keep the examples as they are and make sure that pybids can "see" all the files in it. Will try to expand to identify what are pybids blind spots, this should help.

Reply:
@agahkarakuzu turns that pybids seems OK but that the qsm examples need fixing indeed.


https://github.com/bids-standard/bids-examples/issues/311

Reply:
> Bringing bids-examples on board, 

Seems that importing the bids-examples in the `bids/test/data` is leading to some CI errors

https://github.com/bids-standard/pybids/runs/6067050496?check_suite_focus=true#step:10:967

```
=========================== short test summary info ============================
ERROR bids/tests/data/bids-examples/eeg_matchingpennies/code/format_v016_to_v020.py
ERROR bids/tests/data/bids-examples/eeg_matchingpennies/code/format_v016_to_v020.py
ERROR bids/tests/data/bids-examples/eeg_matchingpennies/code/format_v020_to_v030.py
ERROR bids/tests/data/bids-examples/eeg_matchingpennies/code/format_v020_to_v030.py
```


https://github.com/bids-standard/pybids/runs/6067050496?check_suite_focus=true#step:10:727

```
_ ERROR collecting bids/tests/data/bids-examples/eeg_matchingpennies/code/format_v016_to_v020.py _
bids/tests/data/bids-examples/eeg_matchingpennies/code/format_v016_to_v020.py:28: in <module>
    with open(CHANGES, "r", encoding="utf-8") as fin:
E   FileNotFoundError: [Errno 2] No such file or directory: '/home/stefanappelhoff/Desktop/bids/bids-examples/eeg_matchingpennies/CHANGES'
```

Reply:
> for the record here is an example of what I get for "micr_SPIM"
> 
> ```python
> ds = join(get_test_data_path(), "bids-examples", "micr_SPIM")
> layout = BIDSLayout(ds)
> files = layout.get(return_type="file")
> print(files)
> 
> ['/home/remi/github/pybids/bids/tests/data/bids-examples/micr_SPIM/dataset_description.json', 
> '/home/remi/github/pybids/bids/tests/data/bids-examples/micr_SPIM/participants.json', 
> '/home/remi/github/pybids/bids/tests/data/bids-examples/micr_SPIM/participants.tsv', 
> '/home/remi/github/pybids/bids/tests/data/bids-examples/micr_SPIM/README']
> ```

Hi @Remi-Gau, I tested it as well on a fresh pybids install and I am able to index all the micr files.
With:
```
ds = os.path.join('micr_SPIM')
layout = BIDSLayout(ds)
files = layout.get(return_type="file")
print(*files, sep="\n")
```
I get the 22 expected files:
```
/home/mhbourget/code/pybids/examples/micr_SPIM/dataset_description.json
/home/mhbourget/code/pybids/examples/micr_SPIM/participants.json
/home/mhbourget/code/pybids/examples/micr_SPIM/participants.tsv
/home/mhbourget/code/pybids/examples/micr_SPIM/README
/home/mhbourget/code/pybids/examples/micr_SPIM/samples.json
/home/mhbourget/code/pybids/examples/micr_SPIM/samples.tsv
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-01_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-01_SPIM.ome.tif
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-02_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-02_SPIM.ome.tif
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-03_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-03_SPIM.ome.tif
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-04_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-A_stain-LFB_chunk-04_SPIM.ome.tif
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-01_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-01_SPIM.ome.tif
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-02_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-02_SPIM.ome.tif
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-03_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-03_SPIM.ome.tif
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-04_SPIM.json
/home/mhbourget/code/pybids/examples/micr_SPIM/sub-01/micr/sub-01_sample-B_stain-LFB_chunk-04_SPIM.ome.tif
```
Similarly, `micr_SEM` also index the 12 expected files.

If I'm not mistaken, the BIDSLayout has validate=True by default (with the BIDS-validator). Could it be that your installation and BIDS-validator version don't include microscopy?

I have `bids-validator==1.9.3`

Reply:
Thanks for looking into this.

I checked the exact lines in CI and the tests for micr actually pass, so you are correct I must have messed up something locally.

```
bids/layout/tests/test_layout_on_examples.py::test_layout_on_examples_no_derivatives[micr_SEM-12] 
[gw1] [ 25%] PASSED bids/layout/tests/test_layout_on_examples.py::test_layout_on_examples_no_derivatives[micr_SEM-12] 
bids/layout/tests/test_layout_on_examples.py::test_layout_on_examples_no_derivatives[micr_SPIM-22] 
[gw1] [ 25%] PASSED bids/layout/tests/test_layout_on_examples.py::test_layout_on_examples_no_derivatives[micr_SPIM-22] 
```

Reply:
> Bringing bids-examples on board, could also be a way to start implementing a test of the default path patterns (see [here](https://github.com/bids-standard/pybids/pull/605#issuecomment-626344752) ) until we use the schema generate those (and even then we may want to keep the test) .

Testing path patterns should be done in a different PR